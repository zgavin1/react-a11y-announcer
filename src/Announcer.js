import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Announcer extends Component {

    constructor(props) {
      super(props);
      this.state = {
        text: ''
      }
    }

    static propTypes = {
      text: PropTypes.string,
      politeness: PropTypes.string
    }

    static defaultProps = {
      politeness: 'polite'
    }

    componentWillReceiveProps(nextProps) {
      const currentAnnouncement = this.state.text;
      let nextAnnouncement = nextProps.text;
      if (nextAnnouncement === currentAnnouncement) {
        nextAnnouncement = nextAnnouncement + '\u00A0';
      }
      this.setState(prevState => ({
        text: nextAnnouncement
      }));
    }

    defaultStyles = {
        position: 'absolute',
        visibility: 'visible',
        overflow: 'hidden',
        display: 'block',
        width: '1px',
        height: '1px',
        margin: '-1px',
        border: '0',
        padding: '0',
        clip: 'rect(0px, 0px, 0px, 0px)',
        clipPath: 'polygon(0px 0px, 0px 0px, 0px 0px, 0px 0px)',
        whiteSpace: 'nowrap'
    }

    render() {
        const { className, text, politeness } = this.props;
        const styles = className ? {} : this.defaultStyles;
        return (
            <div
                aria-live={politeness}
                style={styles}
                className={className}
            >
                {
                    this.state.text.length ?
                    <p>{this.state.text}</p> :
                    null
                }
            </div>
        )
    }
}

export default Announcer;
