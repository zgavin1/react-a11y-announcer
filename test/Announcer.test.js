import React from 'react';
import Announcer from '../src/Announcer';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

test('Component is created', () => {
  const wrapper = renderer.create(
    <Announcer text="This is an announcement!" />
  );
  let tree = wrapper.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Should have a default politeness prop of `polite`', () => {
  const wrapper = shallow(<Announcer text="" />);
  const politeness = wrapper.instance().props.politeness;
  expect(politeness).toBe('polite');
});

test('Text node updates when state changes', () => {
  // Since we really can't test the live announcement,
  // we can at least ensure the text node gets updated.
  const wrapper = shallow(<Announcer text="" />);
  const testAnnouncementStr = 'Here\'s a new announcement!';
  expect(wrapper.text()).toEqual('');
  wrapper.setProps({ text: testAnnouncementStr });
  expect(wrapper.text()).toEqual(testAnnouncementStr);
});
