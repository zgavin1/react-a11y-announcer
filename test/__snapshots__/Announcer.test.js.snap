// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is created 1`] = `
<div
  aria-live="polite"
  className={undefined}
  style={
    Object {
      "border": "0",
      "clip": "rect(0px, 0px, 0px, 0px)",
      "clipPath": "polygon(0px 0px, 0px 0px, 0px 0px, 0px 0px)",
      "display": "block",
      "height": "1px",
      "margin": "-1px",
      "overflow": "hidden",
      "padding": "0",
      "position": "absolute",
      "visibility": "visible",
      "whiteSpace": "nowrap",
      "width": "1px",
    }
  }
>
  <p>
    This is an announcement!
  </p>
</div>
`;
